### ISSUE TITLE
Masked Pooling and Convolution

### ISSUE THREAD

**Issue by digital-idiot** [tensorflow/tensorflow#41225]
In geospatial domain often masked images are used due to various reasons. In some specific cases is not sufficient to fill the masked areas as zero and work with that image. Therefore it would be vvery helpful if it would be possible to perform convolution and pooling operations (specifically 2D pooling and convolution), optionally with a mask.

**Comment by ravikyram** [comment]
@digital-idiot 

Do you have any use case that requires the feature you are interested in? Please feel free to submit a PR if you have use cases that supports that feature.Thanks!

**Comment by digital-idiot** [comment]
In general zeroing the masked values works (for example: [PixelRNN](https://arxiv.org/pdf/1601.06759.pdf)) but in some specific use cases if the subsequent pooling layer is average pooling then simply zeroing the values do not work anymore because zeroing the values results into wrong estimate of the mean. Moreover, there is no concrete way to distinguish between a masked pixel and a unmasked pixel with value zero. In many cases normalizing and shifting the value range is not advisable in case of geo-spatial images which further limits the option.

The following 2 paper shows currently how tedious it is achieve masked convolution and pooling are:
https://arxiv.org/pdf/1605.02226v3.pdf
https://arxiv.org/pdf/2006.12486v1.pdf

**Comment by pjpratik** [comment]
@digital-idiot Can you please check [this](https://www.tensorflow.org/guide/keras/masking_and_padding) link for masking. Please let us know if this helps. Thank you.

**Comment by digital-idiot** [comment]
@pjpratik 
`[keras.layers.Masking](https://keras.io/api/layers/core_layers/masking#masking-class)` is supported in a limited number of layers. Unlike `LSTM`, the `Convolution` layers do not have an option to enable masking. Would it be possible to enable masking for `Convolution` layers by wrapping them in a custom layer like the `ReLU` example in your link?

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
